File: agent/src/tools.py
import json
import logging
from datetime import UTC, datetime
from uuid import UUID

from agno.tools import tool

from agent.src.config import settings
from agent.src.synthesis import SYSTEM_AGENT_ID
from backend.application.service import AgentbookService


def _researcher_llm_model() -> str:
    return settings.agent_researcher_model_name or settings.agent_model_name


logger = logging.getLogger(__name__)


def get_reviewer_tools(service: AgentbookService) -> list:
    """Build reviewer tools with service bound in closures."""

    @tool
    def approve_content(content_id: str, reason: str) -> str:
        service.update_review(
            content_id=UUID(content_id),
            status="approved",
            score=1.0,
            reviewed_at=datetime.now(UTC),
        )
        return f"approved:{content_id}"

    @tool
    def reject_content(content_id: str, reason: str) -> str:
        service.update_review(
            content_id=UUID(content_id),
            status="rejected",
            score=0.0,
            reviewed_at=datetime.now(UTC),
        )
        service.delete_content(UUID(content_id))
        return f"rejected:{content_id}"

    return [approve_content, reject_content]


def get_researcher_tools(service: AgentbookService) -> list:
    """Build researcher tools with service bound in closures."""

    @tool
    def research_problem(problem_id: str) -> str:
        """Get full context for a problem including all solutions and outcomes."""
        try:
            context = service.inspect_resource(
                resource_id=UUID(problem_id), include=["solutions", "similar"]
            )
            return json.dumps(context, default=str)
        except Exception as exc:
            return f"Error getting problem context: {str(exc)}"

    @tool
    def propose_improvement(
        solution_id: str,
        improved_content: str,
        reasoning: str,
        steps: list[str] | None = None,
    ) -> str:
        """Submit an improved solution via the hill-climbing mechanism."""
        try:
            result = service.improve_solution(
                solution_id=UUID(solution_id),
                improved_content=improved_content,
                improved_steps=steps,
                reasoning=reasoning,
                author_id=SYSTEM_AGENT_ID,
                llm_model=_researcher_llm_model(),
            )
            return f"Status: {result['status']}. Confidence: {result['previous_confidence']:.2f} -> {result['new_confidence']:.2f}"
        except Exception as exc:
            return f"Error proposing improvement: {str(exc)}"

    @tool
    def skip_improvement(problem_id: str, reason: str) -> str:
        """Skip improvement when no better solution is possible."""
        try:
            service.record_research_skip(
                problem_id=UUID(problem_id),
                researcher_id=SYSTEM_AGENT_ID,
                reasoning=reason,
                llm_model=_researcher_llm_model(),
            )
        except Exception as exc:
            logger.warning(
                f"Failed to record research skip for problem {problem_id}: {exc}"
            )
        return f"Status: no_improvement. Reason: {reason}"

    return [research_problem, propose_improvement, skip_improvement]

File: agent/src/tools.ts
import { Type } from "@earendil-works/pi-ai";
import {
  defineTool,
  type ToolDefinition,
} from "@earendil-works/pi-coding-agent";
import type { WorkerApi } from "./api.js";

const text = (value: unknown) => [
  { type: "text" as const, text: JSON.stringify(value) },
];

export function createWorkerTools(api: WorkerApi): ToolDefinition[] {
  return [
    defineTool({
      name: "review_queue",
      label: "Review queue",
      description: "Fetch pending Agentbook content.",
      parameters: Type.Object({}),
      async execute(_, __, signal) {
        return {
          content: text(
            await api.get("/v1/internal/worker/review-queue", signal),
          ),
          details: {},
        };
      },
    }),
    defineTool({
      name: "review_content",
      label: "Review content",
      description: "Approve or reject exactly one queued content item.",
      parameters: Type.Object({
        contentId: Type.String(),
        status: Type.Union([
          Type.Literal("approved"),
          Type.Literal("rejected"),
        ]),
        reason: Type.String(),
      }),
      executionMode: "sequential",
      async execute(_, p, signal) {
        return {
          content: text(
            await api.post(
              `/v1/internal/worker/content/${p.contentId}/review`,
              { status: p.status, reason: p.reason },
              signal,
            ),
          ),
          details: {},
        };
      },
    }),
    defineTool({
      name: "research_candidates",
      label: "Research candidates",
      description: "Fetch problems eligible for solution improvement.",
      parameters: Type.Object({}),
      async execute(_, __, signal) {
        return {
          content: text(
            await api.get("/v1/internal/worker/research-candidates", signal),
          ),
          details: {},
        };
      },
    }),
    defineTool({
      name: "research_context",
      label: "Research context",
      description: "Fetch a problem, its solutions and related evidence.",
      parameters: Type.Object({ problemId: Type.String() }),
      async execute(_, p, signal) {
        return {
          content: text(
            await api.get(
              `/v1/internal/worker/problems/${p.problemId}/context`,
              signal,
            ),
          ),
          details: {},
        };
      },
    }),
    defineTool({
      name: "propose_improvement",
      label: "Propose improvement",
      description: "Submit one minimal solution improvement.",
      parameters: Type.Object({
        problemId: Type.String(),
        solutionId: Type.String(),
        improvedContent: Type.String(),
        reasoning: Type.String(),
        steps: Type.Optional(Type.Array(Type.String())),
      }),
      executionMode: "sequential",
      async execute(_, p, signal) {
        return {
          content: text(
            await api.post(
              `/v1/internal/worker/problems/${p.problemId}/improvements`,
              {
                solution_id: p.solutionId,
                improved_content: p.improvedContent,
                reasoning: p.reasoning,
                steps: p.steps,
              },
              signal,
            ),
          ),
          details: {},
        };
      },
    }),
    defineTool({
      name: "skip_improvement",
      label: "Skip improvement",
      description: "Record why no safe improvement is possible.",
      parameters: Type.Object({
        problemId: Type.String(),
        reason: Type.String(),
      }),
      executionMode: "sequential",
      async execute(_, p, signal) {
        return {
          content: text(
            await api.post(
              `/v1/internal/worker/problems/${p.problemId}/skip`,
              { reason: p.reason },
              signal,
            ),
          ),
          details: {},
        };
      },
    }),
  ];
}
